File: cpp/27_字符串的排列.cpp
#include <iostream>
#include <vector>
#include <algorithm>

class Solution {
public:
    std::vector<std::string> ret;
    
    void swap(char& a, char& b) {
        char tmp = a;
        a = b;
        b = tmp;
    }
    
    // 采用递归，输入str，每个函数实现的功能是将当前的全排列可能添加到ret中去
    void perm(std::string str, int low, int high) {
        if (low == high) {
            ret.push_back(str);
            return;
        }
        
        sort(str.begin() + low, str.end());
        
        // 依次交换，并进行下一次的迭代
        for (int i = low; i <= high; i++) {
            // 跳过重复
            if (i == low || str[low] != str[i]) {
                swap(str[i], str[low]); // 交换
                perm(str, low + 1, high); // 递归
                //swap(str[i], str[low]); // 交换回来
            }
        }
        return;
    }
    
    std::vector<std::string> Permutation(std::string str) {
        if (str.empty()) {
            return ret;
        }
        
        sort(str.begin(), str.end());
        perm(str, 0, str.size() - 1);
        return ret;
    }
};

int main(int argc, char* argv[])
{
    Solution s;
    std::string ss = "aabc";
    std::vector<std::string> result = s.Permutation(ss);

    for (auto i : result) {
        std::cout << i << std::endl;
    }

    return 0;
}

File: python/27_字符串的排列.py
#!/bin/bash python3
#-*- coding:utf-8 -*-

class Solution:
    def __init__(self):
        self.ret = []
        
    def perm(self, ss, low, high):
        if low == high:
            tmp = "".join(ss)
            self.ret.append(tmp)
            return
        
        # 这里是个坑，需要补上之前的那一小段才可以
        curr = ss[0:low] + sorted(ss[low:high + 1])
        
        for i in range(low, high + 1):
            if i == low or curr[i] != curr[low]:
                curr[i], curr[low] = curr[low], curr[i]
                self.perm(curr, low + 1, high)
                curr[i], curr[low] = curr[low], curr[i]
                
        return
    
    def Permutation(self, ss):
        # write code here
        if ss == None:
            return self.ret
        
        strList = list(ss)
        self.perm(strList, 0, len(strList) - 1)
        
        return self.ret


def main():
    ss = "aabc"
    s = Solution()
    result = s.Permutation(ss)

    print(ss)
    print(result)


if __name__ == "__main__":
    main()